arrows/klv/klv_read_write.txx

```diff
@@ -0,0 +1,346 @@
+T
+klv_read_ber( Iterator& data, size_t max_length )
+{
+  // Ensure types are compatible with our assumptions
+  KLV_ASSERT_UINT( T );
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  if( !max_length )
+  {
+    VITAL_THROW( kwiver::vital::metadata_buffer_overflow,
+                 "BER decoding overruns end of data buffer" );
+  }
+
+  // Short form - first bit is 0, remaining bits are the value itself
+  if( !( 0x80 & *data ) )
+  {
+    return *( data++ );
+  }
+
+  // Long form - first bit is 1, remaining bits are the length of the
+  // following
+  size_t const value_length = ( 0x7F & *data ) + 1;
+
+  if( value_length > max_length )
+  {
+    VITAL_THROW( kwiver::vital::metadata_buffer_overflow,
+                 "BER decoding overruns end of data buffer" );
+  }
+
+  auto const rewind = data++;
+  try
+  {
+    return klv_read_int< T >( data, value_length - 1 );
+  }
+  catch ( const kwiver::vital::metadata_type_overflow& e )
+  {
+    data = rewind;
+
+    throw e;
+  }
+}
+
+// ---------------------------------------------------------------------------
+template < class T, class Iterator >
+void
+klv_write_ber( T value, Iterator& data, size_t max_length )
+{
+  KLV_ASSERT_UINT( T );
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  auto const value_length = klv_ber_length( value );
+  if( value_length > max_length )
+  {
+    VITAL_THROW( kwiver::vital::metadata_buffer_overflow,
+                 "BER encoding overruns end of data buffer" );
+  }
+
+  if( value < 128 )
+  {
+    *( data++ ) = static_cast< uint8_t >( value );
+  }
+  else
+  {
+    *( data++ ) = static_cast< uint8_t >( ( 0x7F & value_length - 1 ) | 0x80 );
+    klv_write_int( value, data, value_length - 1 );
+  }
+}
+
+// ---------------------------------------------------------------------------
+template < class T >
+size_t
+klv_ber_length( T value )
+{
+  KLV_ASSERT_UINT( T );
+  return ( value > 127 ) ? klv_int_length( value ) + 1 : 1;
+}
+
+// ---------------------------------------------------------------------------
+template < class T, class Iterator >
+T
+klv_read_ber_oid( Iterator& data, size_t max_length )
+{
+  KLV_ASSERT_UINT( T );
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  // Highest bit set = not final byte
+  T value = 0;
+  for(; max_length && ( 0x80 & *data ); ++data, --max_length )
+  {
+    if( _left_shift_overflow< T >( value, 7 ) )
+    {
+      VITAL_THROW( kwiver::vital::metadata_type_overflow,
+                   "BER-OID value will overflow given type" );
+    }
+    value <<= 7;
+    value |= 0x7F & *data;
+  }
+
+  if( !max_length )
+  {
+    VITAL_THROW( kwiver::vital::metadata_buffer_overflow,
+                 "BER-OID decoding overruns end of data buffer" );
+  }
+
+  if( _left_shift_overflow< T >( value, 7 ) )
+  {
+    VITAL_THROW( kwiver::vital::metadata_type_overflow,
+                 "BER-OID value will overflow given type" );
+  }
+
+  // Highest bit unset = final byte
+  value <<= 7;
+  value |= *( data++ );
```

Comment:
Somewhat weak suggestion:

```suggestion
  do
  {
    if( !max_length )
    {
      // TODO restore iterator
      VITAL_THROW( kwiver::vital::metadata_buffer_overflow,
                   "BER-OID decoding overruns end of data buffer" );
    }
    --max_length;

    if( _left_shift_overflow< T >( value, 7 ) )
    {
      // TODO restore iterator
      VITAL_THROW( kwiver::vital::metadata_type_overflow,
                   "BER-OID value will overflow given type" );
    }
    value <<= 7;
    value |= 0x7F & *data;
    
    // Highest bit unset = final byte
  } while( 0x80 & *( data++ ) )
```

Please double-check my logic 🙂. Also, there are various ways this could be done (e.g. still using a `for` or using a `while( true )`); the idea is mainly to avoid having separate code paths to add the last byte vs. other bytes.